tensorflow/tools/dist_test/scripts_allreduce/k8s_generate_yaml_lib.py

```diff
@@ -0,0 +1,225 @@
+      volumes: {volumes}
+      - name: dshm
+        emptyDir:
+          medium: Memory
+      - name: sshkeys
+        configMap:
+          name: {config_map}
+          items:
+          - key: publickey
+            path: id_rsa.pub
+          - key: privatekey
+            path: id_rsa
+          - key: sshconfig
+            path: config
+      - name: scripts
+        configMap:
+          name: {config_map}
+          items:
+          - key: start
+            path: start.sh
+""")
+_ENV_VAR_TEMPLATE = '{name: "%s", value: "%s"}'
+
+def GenerateConfig(docker_image,
+                   num_containers,
+                   config_map,
+                   deployment,
+                   port,
+                   use_hostnet,
+                   use_shared_volume,
+                   env_vars=None):
+  """Generate configuration strings.
+
+  Args:
+    docker_image: docker image to use.
+    num_containers: number of containers.
+    config_map: config map.
+    deployment: deployment.
+    port: ssh port.
+    use_hostnet: Used to enable host network mode.
+    use_shared_volume: Used to mount shared volume.
+    env_vars: dictionary of environment variables to set.
+
+  Returns:
+    Kubernetes yaml config.
+  """
+
+  if env_vars is None:
+    env_vars = {}
+  env_str = ', '.join([_ENV_VAR_TEMPLATE % (name, value)
+                       for name, value in env_vars.items()])
+
+  private_key, public_key = generate_RSA(2048)
+
+  CHANGE_SSH_PORT = get_change_ssh_port(use_hostnet, port)
+
+  config = CONFIG_MAP.format(
+      port=port,
+      config_map=config_map,
+      private_key=private_key,
+      public_key=public_key,
+      change_ssh_port=CHANGE_SSH_PORT,
+      env_vars=env_str)
+  config += '---\n\n'
+
+  HOST_NET = get_hostnet(use_hostnet)
+  VOLUME_MOUNTS = get_volume_mounts(use_shared_volume)
+  VOLUMES = get_volumes(use_shared_volume)
+
+  config += DEPLOYMENT.format(
+      deployment=deployment,
+      num_containers=num_containers,
+      docker_image=docker_image,
+      port=port,
+      config_map=config_map,
+      hostnet=HOST_NET,
+      volume_mounts=VOLUME_MOUNTS,
+      volumes=VOLUMES,
+      env_vars=env_str)
+
+  return config
+
+def generate_RSA(bits=2048, exponent=65537):
+  key = RSA.generate(bits, e=exponent)
+  pubkey = key.publickey()
+  
+  private_key = key.exportKey('PEM')
+  public_key = pubkey.exportKey('OpenSSH')
+
+  # Format private_key in yaml file
+  space_before = " " * 4
+  private_key_split = private_key.split('\n')
+  private_key = ''.join(("" if index == 0 else space_before) + line.strip() \
+        + ('\n' if index != len(private_key_split) - 1 else "") \
+        for index, line in enumerate(private_key_split))
+
+  return private_key, public_key
+
+def get_change_ssh_port(use_hostnet, port):
+  if use_hostnet == 1:
+    return "sed -i '/Port 22/c\Port {}' /etc/ssh/sshd_config".format(port)
+
+  return ""
+
+def get_hostnet(use_hostnet):
+  if use_hostnet == 1:
+    return """\n\
```

Comment:
Triple quotes preserve newlines in strings, so I think we can remove `\n\`s?

Reply:
Yes, you are right! We can remove \n\s because triple quotes can span multiple lines and preserve newlines in the strings.